07-routing-gossip.md

```diff
@@ -510,16 +524,197 @@ An endpoint node:
+    - MUST limit `number_of_blocks` to the maximum number of blocks whose
+      results could fit in `encoded_short_ids`
+    - if does not maintain up-to-date channel information for `chain_hash`:
+      - MUST set `complete` to 0.
+    - otherwise:
+      - SHOULD set `complete` to 1.
+
+#### Rationale
+
+A single response might be too large for a single packet, and also a peer can
+store canned results for (say) 1000-block ranges, and simply offer each reply
+which overlaps the ranges of the request.
+
+### The `gossip_timestamp_range` Message
+
+1. type: 265 (`gossip_timestamp_range`) (`gossip_queries`)
+2. data:
+    * [`32`:`chain_hash`]
+    * [`4`:`first_timestamp`]
+    * [`4`:`timestamp_range`]
+
+This message allows a node to constrain future gossip messages to
+a specific range.  A node which wants any gossip messages would have
+to send this, otherwise `gossip_queries` negotiation means no gossip
+messages would be received.
+
+Note that this "filter" replaces any previous one, so it can be used
+multiple times to change the gossip from a peer.
+
+#### Requirements
+
+The sender`:
+  - MUST set `chain_hash` to the 32-byte hash that uniquely identifies the chain
+  that it wants the gossip to refer to.
+
+The receiver:
+  - SHOULD send all gossip messages whose `timestamp` is greater or
+    equal to `first_timestamp`, and less than `first_timestamp` plus
+    `timestamp_range`.
+	- MAY wait for the next outgoing announcement flush to send these.
+    - MUST order these such that `channel_announcement`s precede any corresponding `channel_update`s and `node_announcement`s.
+	- if a `channel_update` has been received for a `channel_announcement`:
+	  - MUST consider the `timestamp` of the `channel_announcement` to be the `timestamp` of a `channel_update`.
+    - otherwise:
+	  - MUST consider the `timestamp` of the `channel_announcement` to be some possible valid creation time.
+  - SHOULD restrict future gossip messages to those whose `timestamp`
+    is greater or equal to `first_timestamp`, and less than
+    `first_timestamp` plus `timestamp_range`.
+
+#### Rationale
+
+Since `channel_announcement` doesn't have a timestamp, we generate
+one.  As `channel_announcement` is usually followed immediately by a
+`channel_update`, that `channel_update` timestamp serves well: in the
+case where no `channel_update` has been received yet, any reasonable
+value will do: it could be the block time of the first possible
+announcement block (ie. six blocks following the funding transaction),
+or simply the this node received it.
+
+This does mean that a new node will consider a `channel_announcement`
+to be more recent than an old node would, but in the common case where
+`gossip_timestamp_range` is sent by the old node to restrict gossip to
+recent messages: in that case `channel_announcement` will be sent
+redundantly.  The new node will already need to ask the old node for
+old gossip, so it will also receive the `channel_announcement`.
+
```

Comment:
There is still a problem with old inactive channels: if you prune and forget a channel, and later receive an announcement for this same channel, but no matching channel updates, you will broadcast again it (unless you set its timestamp to the timestamp of the block it appeared in ?). I think that broadcasting channel announcements only when we have a matching channel update should work. We could also group channel announcements and channel updates, it is less efficient and we'll send more data but it mitigated by the timestamp filter.

Reply:
Is there still an issue with old inactive channels? You'd start the gossip timestamp at the point beyond your "zombie horizon", so the stale channel update would never be sent to you right?

Reply:
Ok, so as agreed in call, timestamp for announce filtering is that of any update. This means you don't send a useless announce which has no channel updates ever. It means you don't have to store the "timestamp" since you can just use the latest update, but you can, and you don't have to deal with the corner case "new update arrived, now announce timestamp is within the window so I have to send it too" since you can argue that you're using the older timestamp. You only need to consider that for the very first update.

Reply:
Yes I think it works (we'll probably go with the "use latest timestamp" approach though). All that is missing now is the `gzip` option :)
We have updated endurance on testnet with a version of eclair that implements this PR

Reply:
I also like the idea of withholding channel announcements until there is a corresponding update to timestamp it, especially compared to the ambiguity of "choose a timestamp that is reasonable." What is the downside to that?